backend/ethereum/channel/register_test.go

```diff
@@ -71,54 +84,69 @@ func registerMultipleConcurrent(t *testing.T, numParts int, parallel bool) {
 	if parallel {
 		wg.Add(numParts)
 	}
+
+	txs := make([]*channel.Transaction, numParts)
+	subs := make([]channel.AdjudicatorSubscription, numParts)
 	for i := 0; i < numParts; i++ {
-		sleepDuration := time.Duration(rng.Int63n(10)+1) * time.Millisecond
-		// manipulate the state
-		state.Version = uint64(int(state.Version) + i)
+		state.Version = uint64(int(state.Version) + i) // manipulate the state
 		tx := testSignState(t, s.Accs, params, state)
+		txs[i] = &tx
+
+		sleepDuration := time.Duration(rng.Int63n(10)+1) * time.Millisecond
 		reg := func(i int, tx channel.Transaction) {
 			if parallel {
 				defer wg.Done()
 				<-startBarrier
 				time.Sleep(sleepDuration)
 			}
+
 			// create subscription
 			adj := s.Adjs[i]
 			sub, err := adj.Subscribe(ctx, params)
 			require.NoError(t, err)
-			defer sub.Close()
+			subs[i] = sub
+
 			// register
 			req := channel.AdjudicatorReq{
 				Params: params,
-				Acc:    s.Accs[i],
-				Idx:    channel.Index(i),
 				Tx:     tx,
 			}
-			assert.NoError(t, adj.Register(ctx, req), "Registering should succeed")
-			event := sub.Next()
-			assert.NotEqual(t, event, &channel.RegisteredEvent{}, "registering should return valid event")
-			assert.False(t, event.Timeout().IsElapsed(ctx),
-				"registering non-final state should return unelapsed timeout")
-			t.Logf("Peer[%d] registered successful", i)
+			err = adj.Register(ctx, req, nil)
+			require.True(t, err == nil || ethchannel.IsErrTxFailed(err), "Registering peer %d", i)
 		}
 		if parallel {
 			go reg(i, tx)
 		} else {
 			reg(i, tx)
 		}
 	}
+
 	if parallel {
 		close(startBarrier)
 		wg.Wait()
 	}
+
+	time.Sleep(100 * time.Millisecond) // Give the subscriptions time to receive the latest event.
```

Comment:
This has to do with the subscription logic again. The idea is that we assume that we always see the latest state. However, on a millisecond basis, this may not be the full truth, as the subscription needs some time to process the events. Hence, we have to wait a bit here.

Do you have a suggestion on what you would like to see here instead?

Reply:
So the next `sub.Next()` call doesn't block until an event is observed?

Reply:
No it currently just returns `nil` :see_no_evil:  
I already fixed it according to #118 since i needed it for the reorg resistance.  
Maybe i can refactor it out and open a MR.

Reply:
The problem here is not that it returns `nil`. I am not even sure if it was doing it here.

The thing is that we want to ensure that we see the events that we have fired above. So if we registered a state with version 10, we check below that we see at least version 10 registered. Because the event subscription is a bit lazy, there is not other way to ensure that with high probability than sleeping a bit and waiting for the subscription to catch up.

Reply:
We should think about how the sub is used in the main code and define a clear expected behavior. E.g., do we want it to return the most recently observed event, and if there isn't any, block until the next is emitted? That's how I remember that it worked before.

What I wanna say is, if we have to do workarounds like this in the tests, it probably means that we should actually fix the sub. But I'm not sure what's the expected behavior on the sub from the core's side.

Reply:
Even though I had a lot of discussions about the adjudicator sub logic with @ggwpez recently and about how I feel it's probably more complex than it needs to be, I don't think this observation here has anything to do with it.

After all, we are not testing the subscription here, we are testing the register function.

The subscription is used to check whether a corresponding registered event has been emitted on the subscription, which needs to have a version >= the registered state version.

It doesn't matter how we implement the subscription. It can always be the case that we don't see an event yet that we just submitted a few milliseconds ago.

So this discussion is more about the implementation of register than the implementation of the subscription. We already created an issue that we should change back register to wait for the corresponding event internally, see #113 .